crates/shared/src/bad_token/cache.rs

```diff
@@ -53,38 +59,43 @@ impl CachingDetector {
         }
     }
 
-    fn insert_into_cache(&self, token: H160, quality: TokenQuality) {
+    async fn insert_into_cache(&self, token: H160, quality: TokenQuality) {
         self.cache
-            .lock()
-            .unwrap()
+            .write()
+            .await
             .insert(token, (Instant::now(), quality));
     }
 
-    fn insert_many_into_cache(&self, tokens: impl Iterator<Item = (H160, TokenQuality)>) {
-        let mut cache = self.cache.lock().unwrap();
+    async fn insert_many_into_cache(&self, tokens: impl Iterator<Item = (H160, TokenQuality)>) {
+        let mut cache = self.cache.write().await;
         let now = Instant::now();
-        for (token, quality) in tokens {
+        tokens.into_iter().for_each(|(token, quality)| {
             cache.insert(token, (now, quality));
-        }
+        });
     }
 
     fn spawn_maintenance_task(self: Arc<Self>) {
-        let cache_expiry = self.cache_expiry;
-        let maintenance_timeout = cache_expiry.div(10).max(Duration::from_secs(60));
+        // We need to prefetch the token quality the `prefetch_time` before the cache
+        // expires
+        let prefetch_time_to_expire = self.cache_expiry - self.prefetch_time;
+        // The maintenance period has to be at least double of the prefetch time in
+        // order to guarantee that the prefetch time is executed before the token
+        // quality expires. This is because of the Nyquist–Shannon sampling theorem.
+        let maintenance_timeout = self.prefetch_time.div(2);
```

Comment:
Shouldn't that be the opposite(according to the comment)?
```suggestion
        let maintenance_timeout = self.prefetch_time.mul(2);
```

Reply:
Good catch! thanks a lot!

Reply:
the comment is incorrect, it is the frequency the one which has to be doubled, not period (which has to be divided by 2)

Reply:
We now running the task every 4 minutes instead of 1. Doesn't that work in the opposite way the theorem states?

Reply:
> the comment is incorrect, it is the frequency the one which has to be doubled, not period (which has to be divided by 2)

Ok, then, shouldn't we use `min(1 minute,  self.prefetch_time.div(2))` then?

Reply:
> So if we change it in the future, we know this, and we don't set a crazily short maintenance_time.

I am worried about the opposite again: too long timeouts. That is why I suggested using the `min` function and not `max.`

Reply:
oh, true, you are right, I got confused because the original code was `max`. Oh yeah, definitely, a really good point! :ok_hand: thanks!

Reply:
ummm actually thinking about that. If the timeouts are too long, it is because the user configured it, and actually running it more often wouldn't have any big benefit, the only benefit is that the token may be updated at the beginning of its `prefetch_time` duration.

We are enforcing that the `cache_expiry` is greater than the `prefetch_time`. If the values are:
- `cache_expiry`: 1 hour
- `prefetch_time`: 30 mins

Then it will run the loop every 15 mins, to check for updates, the cache is valid for 1 hour anyway, so as long as we update it before 1 hour, we are fine. If we run it every minute, it wouldn't have any advantage, just that it may prefetch it sooner, but still, it is prefetch, the token quality isn't out of date, so it is still valid.

So, having the `min` of  minute will just add CPU computation.